llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1419,10 +1419,22 @@ static void narrowToSingleScalarRecipes(VPlan &Plan) {
       // broadcasts.
       if (!vputils::isSingleScalar(RepOrWidenR) ||
           !all_of(RepOrWidenR->users(), [RepOrWidenR](const VPUser *U) {
-            return U->usesScalars(RepOrWidenR) ||
-                   match(cast<VPRecipeBase>(U),
-                         m_CombineOr(m_ExtractLastElement(m_VPValue()),
-                                     m_ExtractLastLanePerPart(m_VPValue())));
+            if (auto *Store = dyn_cast<VPWidenStoreRecipe>(U)) {
+              // The assert must hold as we checked the RepOrWidenR operand
+              // against vputils::isSingleScalar.
+              assert(RepOrWidenR == Store->getAddr() ||
+                     vputils::isSingleScalar(Store->getStoredValue()));
+              return true;
+            }
```

Comment:
Would be good to update the comment above, to explain the store special case. From the tests is looks always beneficial. Wondering if there still could be cases where we create more broadcasts, for something like


```
%a = WIDEN add %x, 1
%b = WIDEN mul %a, 3
store %a
store %b
```

Reply:
Hm, not sure I follow your example: I think you mean that %a's user %b doesn't use scalars of %a, but %b's users use scalars of %b: however, wouldn't we already bail out when checking that %b is a single-scalar (its operand %a isn't single-scalar), and therefore nothing would be narrowed?

Reply:
Oh, right: I missed the reverse-order in my reasoning. Maybe we should work on a follow-up to narrow widen-stores?

Reply:
Hm, see https://godbolt.org/z/3hTncGMfx -- I think a different way to vectorize this would be to insert a broadcast for x? I don't think extra broadcasts will be generated in any case, and that the broadcast will just be moved? There is obviously no diff with the patch, but do you think I should include this test?